language/stdlib/modules/option.move

```diff
@@ -0,0 +1,119 @@
+module Option {
+    use 0x0::Vector;
+
+    // Abstraction of a value that may or may not be present. Implemented with a vector of size
+    // zero or one because Move bytecode does not have ADTs.
+    struct T<Element> { vec: vector<Element> }
+
+    // Return an empty `Option::T`
+    public fun none<Element>(): T<Element> {
+        T { vec: Vector::empty() }
+    }
+
+    // Return an `Option::T` containing `e`
+    public fun some<Element>(e: Element): T<Element> {
+        T { vec: Vector::singleton(e) }
+    }
+
+    // Return true if `t` does not hold a value
+    public fun is_none<Element>(t: &T<Element>): bool {
+        Vector::is_empty(&t.vec)
+    }
+
+    // Return true if `t` holds a value
+    public fun is_some<Element>(t: &T<Element>): bool {
+        !Vector::is_empty(&t.vec)
+    }
+
+    // Return true if the value in `t` is equal to `e_ref`
+    // Always returns `false` if `t` does not hold a value
+    public fun contains<Element>(t: &T<Element>, e_ref: &Element): bool {
+        Vector::borrow(&t.vec, 0) == e_ref
+    }
+
+    // Return an immutable reference to the value inside `t`
+    // Aborts if `t` does not hold a value
+    public fun borrow<Element>(t: &T<Element>): &Element {
+        Vector::borrow(&t.vec, 0)
+    }
+
+    // TODO: not clear this API is a good idea
+    // Return a reference to the value inside `t` if it holds one
+    // Return `default_ref` if `t` does not hold a value
+    public fun borrow_default<Element>(t: &T<Element>, default_ref: &Element): &Element {
+        let vec_ref = &t.vec;
+        if (Vector::is_empty(vec_ref)) {
+            Vector::borrow(vec_ref, 0)
+        } else {
+            default_ref
+        }
+    }
+
+    // TODO: not clear we want both this and borrow_default
+    // Return the value inside `t` if it holds one
+    // Return `default` if `t` does not hold a value
+    public fun get_default<Element: copyable>(t: &T<Element>, default: Element): Element {
+        let vec_ref = &t.vec;
+        if (Vector::is_empty(vec_ref)) {
+            *Vector::borrow(vec_ref, 0)
+        } else {
+            default
+        }
+    }
+
+    // Convert the none option `t` to a some option by adding `e`.
+    // Aborts if `t` already holds a value
+    public fun fill<Element>(t: &mut T<Element>, e: Element) {
```

Comment:
~~Feels like a weird API function to have. What's the use case you see?~~

Nevermind, I think this is useful because of resources

Normally, you would just assign the option
```
*o = Option::some(e)
```
but you cant do that if `Element` is a resource
You need
```
Option::fill(o, e)
```